File: sub-web-service/src/main/java/org/github/web/cache/SysMenuNameJ2Cache.java
package org.github.web.cache;

import java.util.Collection;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Predicate;
import lombok.*;
import org.github.cache.CacheNameSupplier;
import org.github.mysql.web.base.entity.SysMenuMbpEntity;
import org.github.mysql.web.base.service.ISysMenuMbpService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import static com.google.common.collect.ImmutableList.*;
import static com.google.common.collect.ImmutableMap.*;
import static java.util.function.Function.identity;
import static org.github.mysql.web.base.enums.Enable.*;
import static org.github.web.common.CacheName.*;

@Component
public class SysMenuNameJ2Cache extends CustomJ2Cache<SysMenuMbpEntity,String> {
  @Autowired
  private ISysMenuMbpService sysMenuMbpService;

  @Override
  public @NonNull CacheNameSupplier getName() {
    return sysMenuName;
  }

  @Override
  public @NonNull Function<SysMenuMbpEntity,String> getMapper() {
    return SysMenuMbpEntity::getName;
  }

  @Override
  public @NonNull Predicate<SysMenuMbpEntity> getFilter() {
    return v -> v != null && enabled.getCode().equals(v.getEnabled());
  }

  @Override
  public @NonNull Map<String,SysMenuMbpEntity> load(@NonNull Collection<String> keys) {
    val query = sysMenuMbpService.lambdaQuery();
    query.in(!keys.isEmpty(), SysMenuMbpEntity::getMenuId, keys.stream().map(Long::parseLong).collect(toImmutableList()));
    query.select(SysMenuMbpEntity::getMenuId, SysMenuMbpEntity::getName, SysMenuMbpEntity::getEnabled);
    val list = query.list();
    return list.stream().collect(toImmutableMap(v -> v.getMenuId().toString(), identity()));
  }
}

File: sub-web-service/src/main/java/org/github/web/cache/SysMenuNameJ2Cache.kt
package org.github.web.cache

import com.baomidou.mybatisplus.extension.kotlin.KtQueryChainWrapper
import org.github.cache.AbstractJ2Cache
import org.github.cache.CacheNameSupplier
import org.github.mysql.web.base.entity.SysMenuEntity
import org.github.mysql.web.base.enums.Enable.ENABLED
import org.github.mysql.web.base.service.ISysMenuService
import org.github.web.common.CacheName.SYS_MENU_NAME
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.function.Function
import java.util.function.Predicate

@Component
class SysMenuNameJ2Cache : AbstractJ2Cache<SysMenuEntity?, String?>() {
    @Autowired
    private lateinit var sysMenuService: ISysMenuService

    override val name: CacheNameSupplier get() = SYS_MENU_NAME

    override val mapper: Function<SysMenuEntity?, String?> get() = Function { it?.name }

    override val filter: Predicate<SysMenuEntity?> get() = Predicate { it != null && ENABLED.code == it.enabled }

    override fun load(keys: Collection<String>): Map<String, SysMenuEntity?> {
        val query: KtQueryChainWrapper<SysMenuEntity> = sysMenuService.ktQuery()
        query.`in`(keys.isNotEmpty(), SysMenuEntity::menuId, keys.map { it.toLong() })
        query.select(SysMenuEntity::menuId, SysMenuEntity::name, SysMenuEntity::enabled)
        val list: List<SysMenuEntity> = query.list()
        return list.associateBy { it.menuId.toString() }
    }
}
